### PR TITLE
Create Progressive Web App

### PR DESCRIPTION
**Pull request by SohnyBohny**
Create a [PWA](https://developers.google.com/web/progressive-web-apps/) by adding a [app manifest](https://developers.google.com/web/fundamentals/web-app-manifest/) and a [service worker](https://developers.google.com/web/fundamentals/primers/service-workers/)
This will allow users (especially on Android) to add the gitea website to the home-screen and use it like a native app.

Being served over https a service worker will cache all static content which will cause faster page loading and a better user experience.

---

TODOs:
- [x]  Add 192x192 and 512x512 icons to the manifest.json
- [x]  Complete list of static content in serviceworker.js
- [ ] (Reorder head.tmpl)

### CODE REVIEW

**Comment by lafriks** [comment]
Can this be somehow changed not to have duplicate js files? Also if gitea is deployed with suburl this will not work

**Comment by SohnyBohny** [comment]
Cuplicate js files? Where?

What do you mean with suburl? Subdomain should work. But a sub path might cause problems...
The files (manifest and service worker) should be generated dynamically (like html templates) to fix this. But to do that I would need somebody's help.

**Comment by lafriks** [comment]
You coud add `data-cache="true"` to script tags for javascipt etc files that need to be cached and use javascript to select them:
```
var scripts = document.querySelectorAll('script[data-cache="true"]');
// loop over scripts and get src attribute
```
Just an idea, idk if that would work :)

**Comment by SohnyBohny** [comment]
I don't think that this would work because service worker doesn't have access to the document.

**Comment by SohnyBohny** [comment]
The solution might be to generate the service worker file dynamically like html templates. Also the mainfest has a namespace which should be generated dynamically.

**Comment by techknowlogick** [comment]
@SohnyBohny An example of how to use a template to achieve what you are looking for is #3572

**Comment by SohnyBohny** [comment]
Okay thank you - I will check this out when I have time (in ≈2 weeks)

**Comment by 0rzech** [comment]
> What do you mean with suburl? Subdomain should work. But a sub path might cause problems...

@SohnyBohny `AppSubURL` variable stores URL path to Gitea instance [[source](/go-gitea/gitea/blob/b82c14b3d2259912b47fa292b85772ba1d2493d0/modules/setting/setting.go#L768)]. In templates its name is `AppSubUrl` [[source](/go-gitea/gitea/blob/6e64f9db8eb889f9cc7e8c9576b2f9c89750927e/modules/templates/helper.go#L45)].

**Comment by lunny** [comment]
CI failed

**Comment by SohnyBohny** [comment]
What does this error mean? 
``` Makefile:152: recipe for target 'fmt-check' failed ```

**Comment by lunny** [comment]
your golang code is not the same as the format program did.

**Comment by SohnyBohny** [comment]
Finally CI build has passed :-)

File: templates/base/head.tmpl [hunk 1]
```diff
@@ -5,6 +5,23 @@
 	<meta name="viewport" content="width=device-width, initial-scale=1">
 	<meta http-equiv="x-ua-compatible" content="ie=edge">
 	<title>{{if .Title}}{{.Title}} - {{end}}{{AppName}}</title>
+	<link rel="manifest" href="/manifest.json">
+	
+	<script>
+		if ('serviceWorker' in navigator) {
+  			window.addEventListener('load', function() {
+    			navigator.serviceWorker.register('/serviceworker.js').then(function(registration) {
```
Comment (techknowlogick) [inline templates/base/head.tmpl]: `/serviceworker.js` probably needs to be prefaced by `{{AppSubUrl}}`

**Comment by SohnyBohny** [comment]
The 192x192 and 512x512 icons are needed that chrome automatically shows an [install banner](https://developers.google.com/web/fundamentals/app-install-banners/) on Android. Without them the User will not find out this feature.
I'm not sure how these icon are added. I have found an `generate-images` section in the Makefile but I'm not experienced with Makefiles at all.

File: templates/pwa/serviceworker_js.tmpl [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  '{{AppSubUrl}}/vendor/plugins/dropzone/dropzone.css',
+  '/css/theme-{{DefaultTheme}}.css',
+  
+  // img
+  '{{AppSubUrl}}/img/gitea-sm.png',
+  '{{AppSubUrl}}/img/gitea-lg.png',
+
+  // fonts
+  '{{AppSubUrl}}/vendor/plugins/semantic/themes/default/assets/fonts/icons.woff2',
+  '{{AppSubUrl}}/vendor/assets/octicons/octicons.woff2', // TODO: ?MD5
```
Comment (SohnyBohny) [inline templates/pwa/serviceworker_js.tmpl]: The actual requested URL is `/vendor/assets/octicons/octicons.woff2?ef21c39f0ca9b1b5116e5eb7ac5eabe6`. For me it looks like a `MD5` hash is added to the URL. How can this be accessed from the template?

**Comment by lunny** [comment]
~But cannot work on chrome macOS? I have enabled PWA desktop but it seems it doesn't work.~

**Comment by SohnyBohny** [comment]
@lunny any error message? I think 512px sized icon is needed...

**Comment by SohnyBohny** [comment]
### :iphone: On Android:
screenshot_20181030-134129
---
### :desktop_computer: In Chromium Browser (Linux):
screenshot
---
:red_circle: **Note: must be served over https or localhost**

File: templates/pwa/serviceworker_js.tmpl (hunk 2 above)
Comment (techknowlogick) [reply templates/pwa/serviceworker_js.tmpl]: I think fonts may need to have hardcoded MD5s in the URL as the minified css that is provided from octicons has the URL already in there.

**Comment by lunny** [comment]
```
[Macaron] 2018-11-19 09:28:21: Started GET /css/theme-gitea.css for [::1]
2018/11/19 09:28:21 [D] Session ID: 529e98b5f23de857
2018/11/19 09:28:21 [D] CSRF Token: MWPHepbatiwdXEvp0ukobEJxZTU6MTU0MjU5MDg2OTM1NTAwNjAwMA==
2018/11/19 09:28:21 [I] [SQL] SELECT  TOP 1 "id", "lower_name", "name", "full_name", "email", "keep_email_private", "passwd", "must_change_password", "login_type", "login_source", "login_name", "type", "location", "website", "rands", "salt", "language", "created_unix", "updated_unix", "last_login_unix", "last_repo_visibility", "max_repo_creation", "is_active", "is_admin", "allow_git_hook", "allow_import_local", "allow_create_organization", "prohibit_login", "avatar", "avatar_email", "use_custom_avatar", "num_followers", "num_following", "num_stars", "num_repos", "description", "num_teams", "num_members", "diff_view_style" FROM "user" WHERE "lower_name"=? []interface {}{"css"}
2018/11/19 09:28:21 [D] Template: status/404
[Macaron] 2018-11-19 09:28:21: Completed GET /css/theme-gitea.css 404 Not Found in 100.550709ms
```

File: templates/pwa/serviceworker_js.tmpl [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  '{{AppSubUrl}}/vendor/plugins/simplemde/simplemde.min.css',
+  '{{AppSubUrl}}/vendor/plugins/gitgraph/gitgraph.css',
+  '{{AppSubUrl}}/vendor/plugins/tribute/tribute.css',
+  '{{AppSubUrl}}/vendor/plugins/semantic/semantic.min.css',
+  '{{AppSubUrl}}/css/index.css?v={{MD5 AppVer}}',
+  '{{AppSubUrl}}/vendor/plugins/highlight/github.css',
+  '{{AppSubUrl}}/vendor/plugins/jquery.minicolors/jquery.minicolors.css',
+  '{{AppSubUrl}}/vendor/plugins/jquery.datetimepicker/jquery.datetimepicker.css',
+  '{{AppSubUrl}}/vendor/plugins/dropzone/dropzone.css',
+  '/css/theme-{{DefaultTheme}}.css',
```
Comment (techknowlogick) [inline templates/pwa/serviceworker_js.tmpl]:
```
{{if ne DefaultTheme "gitea"}}
 	'{{AppSubUrl}}/css/theme-{{DefaultTheme}}.css',
{{end}}
```

cc: @lunny

Comment (SohnyBohny) [reply templates/pwa/serviceworker_js.tmpl]: Of course feel free to extend the list at any time :smile::+1:

**Comment by lunny** [comment]
Another two TODOs?

File: templates/pwa/serviceworker_js.tmpl (hunk 3 above)
Comment (techknowlogick) [reply templates/pwa/serviceworker_js.tmpl]: @SohnyBohny thanks. I've added that in. @lunny the error you have shouldn't happen anymore, please try again.

**Comment by techknowlogick** [comment]
@lunny I think last TODO can wait for another PR.

**Review by techknowlogick — APPROVED** [review]

**Comment by SohnyBohny** [comment]
Great @techknowlogick thank you

Of course the last TODO can wait (if it is neccessary at all). I was just not sure how you like the head of the html to be ordered.

For me it's done now. Of course the list should be up to date at any time.

Somewere in the docs or in the release notes there should be a note that the service worker has to be served over https (or localhost)

**Review by lunny — APPROVED** [review]

**Comment by lafriks** [comment]
I think this list is already outdated :) We should really think of some kind of automated sync for it

**Comment by lunny** [comment]
Which list?

**Comment by lafriks** [comment]
I think at least recently merged vue-calendar-heatmap, moment

**Comment by SohnyBohny** [comment]
[Workbox](https://developers.google.com/web/tools/workbox/guides/precache-files/cli) is exactly our usecase. I'll look into that as soon as possible and will create a new PR then.

**Comment by SohnyBohny** [comment]
---
Is this the right place to keep discussing about this topic or should I open an issue for that?

---
Anyway:
I've looked into [Workbox](https://developers.google.com/web/tools/workbox/guides/precache-files/) now. The way it works is by looking for specific file types in specified folders and automatically add them to the caching list while building time.

Most cacheable files are located under `/vendor/plugins/` but they are not accessible in [source folder](https://github.com/go-gitea/gitea/tree/master/vendor). Could somebody please explain how this works or link me to a wiki page?

More specific I need following questions answered:
```
- What is the root of your web app (i.e. which directory do you deploy)?
- Which file types would you like to precache? (Located in the folder specified above or in subfolders)
```

**Pull request merged**